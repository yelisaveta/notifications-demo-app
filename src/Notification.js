import React, { Component } from "react";
import PropTypes from "prop-types";
import Card from "@material-ui/core/Card";
import CardActions from "@material-ui/core/CardActions";
import CardContent from "@material-ui/core/CardContent";
import Button from "@material-ui/core/Button";
import DeleteIcon from "@material-ui/icons/Delete";
import { withStyles } from "@material-ui/core/styles";

import LikeButton from "./LikeButton";
import NotificationHeader from "./NotificationHeader";

const styles = theme => ({
  card: {
    width: "100%"
  },
  rightIcon: {
    marginLeft: theme.spacing.unit
  }
});

export class Notification extends Component {
  handleDelete = id => {
    return () => {
      this.props.onDelete(id);
    };
  };

  handleLike = id => {
    return () => {
      this.props.onLike(id);
    };
  };

  render() {
    const { id, title, message, liked, classes } = this.props;
    return (
      <Card className={classes.card} data-testid={`notification-id-${id}`}>
        <NotificationHeader title={title} liked={liked} />
        <CardContent>{message}</CardContent>
        <CardActions>
          <LikeButton onLike={this.handleLike(id)} />
          <Button
            variant="contained"
            color="primary"
            size="medium"
            onClick={this.handleDelete(id)}
          >
            Delete
            <DeleteIcon className={classes.rightIcon} />
          </Button>
        </CardActions>
      </Card>
    );
  }
}

Notification.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  message: PropTypes.string.isRequired,
  liked: PropTypes.bool,
  onDelete: PropTypes.func.isRequired,
  onLike: PropTypes.func.isRequired,
  classes: PropTypes.shape({
    card: PropTypes.string.isRequired,
    rightIcon: PropTypes.string.isRequired
  })
};

export default withStyles(styles)(Notification);
