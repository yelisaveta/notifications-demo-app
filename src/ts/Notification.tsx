import * as React from "react";
import Card from "@material-ui/core/Card";
import CardActions from "@material-ui/core/CardActions";
import CardContent from "@material-ui/core/CardContent";
import Button from "@material-ui/core/Button";
import DeleteIcon from "@material-ui/icons/Delete";
import { withStyles } from "@material-ui/core/styles";
import { Theme } from "@material-ui/core/styles/createMuiTheme";

import LikeButton from "../LikeButton";
import NotificationHeader from "../NotificationHeader";

export interface NotificationProps {
  id: string;
  title: string;
  message: string;
  liked: boolean;
  onDelete: Function;
  onLike: Function;
  classes: {
    card: string;
    rightIcon: string;
  };
}

const styles = (theme: Theme) => ({
  card: {
    width: "100%"
  },
  rightIcon: {
    marginLeft: theme.spacing.unit,
    marginRight: 20
  }
});

export class Notification extends React.Component<NotificationProps, {}> {
  handleDelete = (id: string) => {
    return () => {
      this.props.onDelete(id);
    };
  };

  handleLike = (id: string) => {
    return () => {
      this.props.onLike(id);
    };
  };

  render() {
    const { id, title, message, liked, classes } = this.props;
    return (
      <Card className={classes.card} data-testid={`notification-id-${id}`}>
        <NotificationHeader title={title} liked={liked} />
        <CardContent>{message}</CardContent>
        <CardActions>
          <LikeButton onLike={this.handleLike(id)} />
          <Button
            variant="contained"
            color="primary"
            size="small"
            onClick={this.handleDelete(id)}
          >
            Delete
            <DeleteIcon className={classes.rightIcon} />
          </Button>
        </CardActions>
      </Card>
    );
  }
}

export default withStyles(styles)(Notification);
