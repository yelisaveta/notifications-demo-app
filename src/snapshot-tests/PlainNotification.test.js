import React from "react";
import renderer from "react-test-renderer";

// import plain React component class instead of the HOC (withStyles)
import { Notification as PlainNotification } from "../Notification";

const props = {
  id: "notification-test-id",
  title: "my test title",
  message: "my test message",
  onDelete: jest.fn(),
  onLike: jest.fn()
};

// provide mock class names since withStyles HOC no longer does it
const classes = {
  card: "mock-card-classname",
  rightIcon: "mock-rightIcon-classname"
};

/**
 * Might be a bit better since it reduces dependency on MUI class names.
 * Underlying components still create MUI class names dynamically and causing tests to fail when class names change.
 * Probably a sign that we don't need snapshot test at this level.
 */
describe("PlainNotification without MUI", () => {
  it("component renders correctly", () => {
    const tree = renderer
      .create(<PlainNotification {...props} classes={classes} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
