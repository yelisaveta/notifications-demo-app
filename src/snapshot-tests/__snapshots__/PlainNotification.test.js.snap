// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlainNotification without MUI component renders correctly 1`] = `
<div
  className="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 MuiCard-root-1 mock-card-classname"
  data-testid="notification-id-notification-test-id"
>
  <div
    className="MuiCardHeader-root-29"
  >
    <div
      className="MuiCardHeader-content-32"
    >
      <span
        className="MuiTypography-root-35 MuiTypography-headline-40 MuiCardHeader-title-33"
      >
        my test title
      </span>
    </div>
  </div>
  <div
    className="MuiCardContent-root-71"
  >
    my test message
  </div>
  <div
    className="MuiCardActions-root-72"
  >
    <button
      className="MuiButtonBase-root-101 MuiButton-root-75 MuiButton-outlined-83"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-76"
      >
        Like
      </span>
      <span
        className="MuiTouchRipple-root-113"
      />
    </button>
    <button
      className="MuiButtonBase-root-101 MuiButton-root-75 MuiButton-contained-86 MuiButton-containedPrimary-87 MuiButton-raised-89 MuiButton-raisedPrimary-90 MuiButton-sizeSmall-98 MuiCardActions-action-74"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-76"
      >
        Delete
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-104 mock-rightIcon-classname"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
        </svg>
      </span>
      <span
        className="MuiTouchRipple-root-113"
      />
    </button>
  </div>
</div>
`;
